bd-resilient-kv/src/kv_journal/in_memory.rs

```diff
@@ -0,0 +1,446 @@
+
+    // Write metadata object
+    encode_into_buf(&mut cursor, &Value::Object(metadata))
+      .map_err(|e| anyhow::anyhow!("Failed to encode metadata object: {e:?}"))?;
+
+    // Return new current position (17 + bytes written)
+    Ok(17 + (buffer_len - 17 - cursor.remaining_mut()))
+  }
+
+  /// Get current timestamp in nanoseconds since UNIX epoch.
+  fn current_timestamp() -> anyhow::Result<u64> {
+    SystemTime::now()
+      .duration_since(UNIX_EPOCH)
+      .map_err(|e| anyhow::anyhow!("System time error: {e}"))
+      .and_then(|d| {
+        u64::try_from(d.as_nanos()).map_err(|e| anyhow::anyhow!("Timestamp overflow: {e}"))
+      })
+  }
+
+  fn set_position(&mut self, position: usize) {
+    self.position = position;
+    let position_bytes = (position as u64).to_le_bytes();
+    self.buffer[8 .. 16].copy_from_slice(&position_bytes);
+
+    // Check high water mark
+    self.check_high_water_mark();
+  }
+
+  /// Check if the current position has exceeded the high water mark and trigger callback if needed.
+  fn check_high_water_mark(&mut self) {
+    if !self.high_water_mark_triggered && self.position >= self.high_water_mark {
+      self.high_water_mark_triggered = true;
+      if let Some(callback) = self.high_water_mark_callback {
+        callback(self.position, self.buffer.len(), self.high_water_mark);
+      }
+    }
+  }
+
+  /// Extract the initialization timestamp from a buffer containing journal data.
+  fn extract_timestamp_from_buffer(buffer: &[u8]) -> anyhow::Result<u64> {
+    // We need to create a temporary closed array to parse
+    let mut temp_buffer = buffer.to_vec();
+    temp_buffer.push(ContainerEnd as u8); // Close off the open array
+
+    let (_, decoded) = from_slice(&temp_buffer)
+      .map_err(|e| anyhow::anyhow!("Failed to decode buffer for timestamp extraction: {e:?}"))?;
+
+    if let Value::Array(entries) = decoded {
+      for entry in &entries {
+        if let Value::Object(obj) = entry
+          && obj.contains_key("initialized")
+        {
+          if let Some(Value::Unsigned(timestamp)) = obj.get("initialized") {
+            return Ok(*timestamp);
+          } else if let Some(Value::Signed(timestamp)) = obj.get("initialized") {
+            // Handle the case where timestamp was stored as signed
+            #[allow(clippy::cast_sign_loss)]
+            return Ok(*timestamp as u64);
+          }
+        }
+      }
+    }
+
+    anyhow::bail!("No valid metadata with initialized timestamp found");
+  }
+
+  fn write_journal_entry(&mut self, key: &str, value: &Value) -> anyhow::Result<()> {
+    let initial_position = self.position;
+    let buffer_len = self.buffer.len();
+    let mut cursor = &mut self.buffer[self.position ..];
+
+    // Write the journal entry using BufMut - position is tracked automatically
+    serialize_map_begin(&mut cursor)
+      .map_err(|e| anyhow::anyhow!("Failed to serialize map begin: {e:?}"))?;
```

Comment:
nit: If some of these practically cannot fail I would use the InvariantError to be more clear and also save binary space as these error strings are going to wind up in the binary. Same elsewhere.